src/cockpit/polkit.py

```diff
@@ -100,56 +100,52 @@ async def _communicate(self, process: asyncio.subprocess.Process):
+    """Register polkit agent when required
 
-    Returns an opaque object which needs to be kept during the lifetime of the agent.
-    After that, pass it to @unregister_agent().
+    Use this as a context manager to ensure that the agent gets unregistered again.
     """
-    try:
-        system_bus = bus.Bus.default_system()
-    except OSError as e:
-        logger.debug('cannot connect to system bus, not registering polkit agent: %s', e)
-        return None
-
-    try:
-        # may refine that with a D-Bus call to logind
-        subject = ('unix-session', {'session-id': {'t': 's', 'v': os.environ['XDG_SESSION_ID']}})
-    except KeyError:
-        logger.debug('XDG_SESSION_ID not set, not registering polkit agent')
-        return None
-
-    agent_object = org_freedesktop_PolicyKit1_AuthenticationAgent(responder)
-    agent_slot = system_bus.add_object('/', agent_object)
-
-    # register agent
-    locale_name = locale.setlocale(locale.LC_MESSAGES, '')
-    await system_bus.call_method_async(
-        'org.freedesktop.PolicyKit1',
-        '/org/freedesktop/PolicyKit1/Authority',
-        'org.freedesktop.PolicyKit1.Authority',
-        'RegisterAuthenticationAgent',
-        '(sa{sv})ss',
-        subject, locale_name, '/')
-    logger.debug('Registered agent for %r and locale %s', subject, locale_name)
-    return {'subject': subject, 'slot': agent_slot}
-
-
-async def unregister_agent(agent):
-    """Unregister polkit agent.
-
-    You have to do this explicitly (not just rely on garbage collection), otherwise
-    the next register_agent() will fail with "there is already a registered agent".
-
-    This receives the agent object returned by @register_agent().
-    """
-    system_bus = bus.Bus.default_system()
-    await system_bus.call_method_async(
-        'org.freedesktop.PolicyKit1',
-        '/org/freedesktop/PolicyKit1/Authority',
-        'org.freedesktop.PolicyKit1.Authority',
-        'UnregisterAuthenticationAgent',
-        '(sa{sv})s',
-        agent['subject'], '/')
-    agent['slot'].cancel()
-    logger.debug('Unregistered agent for %r', agent['subject'])
+    def __init__(self, responder: InteractionResponder):
+        self.responder = responder
+        self.agent_slot = None
+
+    async def __aenter__(self):
+        try:
+            self.system_bus = bus.Bus.default_system()
+        except OSError as e:
+            logger.debug('cannot connect to system bus, not registering polkit agent: %s', e)
+            return self
+
+        try:
+            # may refine that with a D-Bus call to logind
+            self.subject = ('unix-session', {'session-id': {'t': 's', 'v': os.environ['XDG_SESSION_ID']}})
+        except KeyError:
+            logger.debug('XDG_SESSION_ID not set, not registering polkit agent')
+            return self
+
+        agent_object = org_freedesktop_PolicyKit1_AuthenticationAgent(self.responder)
+        self.agent_slot = self.system_bus.add_object('/', agent_object)
```

Comment:
Can you export that at a more descriptive path?

Reply:
I never really understood the common practice of D-Bus to have excessive object paths -- '/' is just fine. I introduced an `AGENT_DBUS_PATH` constant and changed it to `/agent`.